modules/gmake2/gmake2_cpp.lua

```diff
@@ -163,31 +163,19 @@
 
 		-- process custom build commands.
 		if fileconfig.hasCustomBuildRule(filecfg) then
-			local env = table.shallowcopy(filecfg.environ)
-			env.PathVars = {
-				["file.basename"]     = { absolute = false, token = node.basename },
-				["file.abspath"]      = { absolute = true,  token = node.abspath },
-				["file.relpath"]      = { absolute = false, token = node.relpath },
-				["file.name"]         = { absolute = false, token = node.name },
-				["file.objname"]      = { absolute = false, token = node.objname },
-				["file.path"]         = { absolute = true,  token = node.path },
-				["file.directory"]    = { absolute = true,  token = path.getdirectory(node.abspath) },
-				["file.reldirectory"] = { absolute = false, token = path.getdirectory(node.relpath) },
-			}
-
-			local shadowContext = p.context.extent(filecfg, env)
```

Comment:
It's probably better to fix this than to remove it. `%{file.directory}` and `%{file.abspath}` should give an absolute path, but won't without the shadow context changes above - the `env.PathVars` changing to `env.pathVars` will fix that.

Reply:
Not sure what `p.context.extent` does, but I go from `#filecfg.buildOutputs == 1` to `#shadowContext.buildOutputs == 0`.

Other generators don't have those... (and indeed absolute path is not respected :/ )